src/main/java/blusunrize/immersiveengineering/common/crafting/SpeedloaderRecipe.java

```diff
@@ -8,69 +8,130 @@
-import net.minecraft.item.crafting.SpecialRecipeSerializer;
-import net.minecraft.util.IItemProvider;
+import net.minecraft.item.crafting.SpecialRecipe;
 import net.minecraft.util.NonNullList;
 import net.minecraft.util.ResourceLocation;
+import net.minecraft.world.World;
 
 import javax.annotation.Nonnull;
 
-public class SpeedloaderRecipe extends ShapedRecipe
+public class SpeedloaderRecipe extends SpecialRecipe
 {
-	public static final IRecipeSerializer<SpeedloaderRecipe> SERIALIZER = IRecipeSerializer.register(
-			ImmersiveEngineering.MODID+":speedloader", new SpecialRecipeSerializer<>(SpeedloaderRecipe::new)
-	);
+	private final byte[] offsetPattern = {0, 1, 1, 1, 0, -1, -1, -1};
 
-	public SpeedloaderRecipe(ResourceLocation id)
+	public SpeedloaderRecipe(ResourceLocation resourceLocation)
 	{
-		super(id, null, 3, 3, getPattern(), new ItemStack(Weapons.speedloader));
+		super(resourceLocation);
 	}
 
-	private static NonNullList<Ingredient> getPattern()
+	@Override
+	public boolean matches(CraftingInventory inv, @Nonnull World world)
 	{
-		Ingredient bullet = Ingredient.fromItems(Weapons.bullets.values().toArray(new IItemProvider[0]));
-		Ingredient speedloader = Ingredient.fromItems(Weapons.speedloader);
-		return NonNullList.from(Ingredient.EMPTY,
-				bullet, bullet, bullet,
-				bullet, speedloader, bullet,
-				bullet, bullet, bullet);
+		ItemStack stackInSlot;
+		int speedloaderX = -1;
+		int speedloaderY = -1;
+		boolean hasSpeedloader = false;
+		boolean hasBullets = false;
+		int width = inv.getWidth();
+		//for bullets found before the speedloader was located. could maybe limit it to half the crafting grids size,
+		//but I can't come up with a smart way to prevent out-of-bounds exceptions for mismatching inventories at the moment
+		boolean[][] blindBulletGrid = new boolean[inv.getHeight()][width];
+		for(int i = 0; i < inv.getSizeInventory(); i++)
+		{
+			stackInSlot = inv.getStackInSlot(i);
+			if(!stackInSlot.isEmpty())
+			{
+				if(stackInSlot.getItem() instanceof SpeedloaderItem)
+				{
+					if(hasSpeedloader||!((SpeedloaderItem)stackInSlot.getItem()).isEmpty(stackInSlot))
+						return false;
+					speedloaderX = i%width;
+					speedloaderY = i/width;
+					for(int j = 0; j < inv.getHeight(); j++)
+					{
+						for(int k = 0; k < width; k++)
+						{
+							if(j >= speedloaderY&&k >= speedloaderX)
+								break;
+							if(blindBulletGrid[j][k]&&(Math.abs(j-speedloaderY) > 1||Math.abs(k-speedloaderX) > 1))
+								return false;
+						}
+						if(j >= speedloaderY)
+							break;
+					}
+					hasSpeedloader = true;
+				}
+				else if(stackInSlot.getItem() instanceof BulletItem)
+				{
+					hasBullets = true;
+					if(!hasSpeedloader)
+						blindBulletGrid[i/width][i%width] = true;
+					else if(Math.abs((i/width)-speedloaderY) > 1||Math.abs((i%width)-speedloaderX) > 1)
+						return false;
+				}
+				else
+					return false;
+			}
+		}
+		return hasSpeedloader&&hasBullets;
```

Comment:
This implementation seems unnecessarily complicated. I'd suggest first finding the speedloader and then checking the bullets. Yes, it will (theoretically) be slower, but there's only one of these recipes, so performance should not be an issue.

Reply:
replaced it with a more necessarily complicated check that now allows for adding bullets to a non-empty speedloader :3